auth_service/auth_adapter/core/auth.py

```diff
@@ -18,121 +18,207 @@
+            self.external_algs = external_algs
+        else:
+            log.warning("Allowed external signing algorithms not configured.")
+        authority_url = config.oidc_authority_url
+        if authority_url:
+            self.check_claims["iss"] = authority_url
+        else:
+            log.warning("No OIDC authority URL configured.")
+        client_id = config.oidc_client_id
+        if client_id:
+            self.check_claims["client_id"] = client_id
+        else:
+            log.warning("No OIDC client ID configured.")
+
+
+jwt_config = JWTConfig()
+
+
+def _compare_user(user: User, external_claims: dict[str, Any]) -> Optional[str]:
+    """Compare user with external claims and return inactivation context."""
+    if user.status == UserStatus.ACTIVATED:
+        if user.name != external_claims.get("name"):
+            return "name change"
+        if user.email != external_claims.get("email"):
+            return "email change"
+    return None
 
 
-def exchange_token(external_token: Optional[str]) -> Optional[str]:
+def _get_inactivated_user(user: User, context: str) -> User:
+    """Get an inactivated copy of the User object."""
+    return user.copy(
+        update=dict(
+            status=UserStatus.INACTIVATED.value,
+            status_change=StatusChange(
+                previous=user.status,
+                by=None,
+                context=context,
+                change_date=datetime.now(),
+            ),
+        )
+    )
+
+
+async def exchange_token(
+    external_token: Optional[str], pass_sub: bool = False, user_dao: UserDao = Depends()
+) -> Optional[str]:
     """Exchange the external token against an internal token.
 
     If the provided external token is valid, a corresponding internal token
     will be returned. Otherwise None will be returned.
+
+    The internal token will contain the name and email taken from the
+    external token, and also its issued date and expiry date.
+    If the user is already registered, the user id and status will be
+    included in the internal token as well.
+    If name or email do not match with the external token, the user
+    is automatically deactivated in the database and internal token.
+    If the user is not yet registered, and pass_sub is set, then the sub claim
+    will be included in the internal token as "ls_id", otherwise and empty string
+    will be returned instead of the internal token.
     """
-    payload = decode_and_verify_token(external_token)
-    if not payload or not isinstance(payload, dict):
-        return None
-    name = payload.get("name")
-    email = payload.get("email")
-    if not (name and email):
-        log.debug("Access token does not contain required user info")
-        return None
-    payload = dict(name=name, email=email)
-    internal_token = sign_and_encode_token(payload)
+    external_claims = decode_and_validate_token(external_token)
+    if not external_claims:
+        return None  # invalid token (error)
```

Comment:
Maybe also here we can simply raise a exception for all the cases that lead to an invalid error.
If needed we can catch and further handle that exception upstream in the call stack.